Python/jit.c

```diff
@@ -0,0 +1,363 @@
+            assert(IS_AARCH64_MOV(*loc32));
+            // Check the implicit shift (this is "part 2 of 3"):
+            assert(get_bits(*loc32, 21, 2) == 2);
+            set_bits(loc32, 5, 16, value, 32);
+            return;
+        case HoleKind_R_AARCH64_MOVW_UABS_G3:
+            // 16-bit high part of an absolute address.
+            assert(IS_AARCH64_MOV(*loc32));
+            // Check the implicit shift (this is "part 3 of 3"):
+            assert(get_bits(*loc32, 21, 2) == 3);
+            set_bits(loc32, 5, 16, value, 48);
+            return;
+        case HoleKind_ARM64_RELOC_GOT_LOAD_PAGE21:
+            // 21-bit count of pages between this page and an absolute address's
+            // page... I know, I know, it's weird. Pairs nicely with
+            // ARM64_RELOC_GOT_LOAD_PAGEOFF12 (below).
+            assert(IS_AARCH64_ADRP(*loc32));
+            // Number of pages between this page and the value's page:
+            value = (value >> 12) - ((uint64_t)location >> 12);
+            // Check that we're not out of range of 21 signed bits:
+            assert((int64_t)value >= -(1 << 20));
+            assert((int64_t)value < (1 << 20));
+            // value[0:2] goes in loc[29:31]:
+            set_bits(loc32, 29, 2, value, 0);
+            // value[2:21] goes in loc[5:26]:
+            set_bits(loc32, 5, 19, value, 2);
+            return;
+        case HoleKind_ARM64_RELOC_GOT_LOAD_PAGEOFF12:
+            // 12-bit low part of an absolute address. Pairs nicely with
+            // ARM64_RELOC_GOT_LOAD_PAGE21 (above).
+            assert(IS_AARCH64_LDR_OR_STR(*loc32) || IS_AARCH64_ADD_OR_SUB(*loc32));
+            // There might be an implicit shift encoded in the instruction:
+            uint8_t shift = 0;
+            if (IS_AARCH64_LDR_OR_STR(*loc32)) {
+                shift = (uint8_t)get_bits(*loc32, 30, 2);
+                // If both of these are set, the shift is supposed to be 4.
+                // That's pretty weird, and it's never actually been observed...
+                assert(get_bits(*loc32, 23, 1) == 0 || get_bits(*loc32, 26, 1) == 0);
+            }
+            value = get_bits(value, 0, 12);
+            assert(get_bits(value, 0, shift) == 0);
+            set_bits(loc32, 10, 12, value, shift);
+            return;
+    }
+    Py_UNREACHABLE();
+}
+
+static void
+copy_and_patch(char *base, const Stencil *stencil, uint64_t *patches)
+{
+    memcpy(base, stencil->body, stencil->body_size);
+    for (uint64_t i = 0; i < stencil->holes_size; i++) {
+        patch(base, &stencil->holes[i], patches);
+    }
+}
+
+static void
+emit(const StencilGroup *group, uint64_t patches[])
+{
+    copy_and_patch((char *)patches[HoleValue_CODE], &group->code, patches);
+    copy_and_patch((char *)patches[HoleValue_DATA], &group->data, patches);
+}
+
+// Compiles executor in-place. Don't forget to call _PyJIT_Free later!
+int
+_PyJIT_Compile(_PyExecutorObject *executor)
```

Comment:
Can't just return the memory, since we need to return the size too. Since we'd probably be returning a struct or using 1-2 outparams, I think just working with the executor itself is a bit simpler.

Also, how would this make `make_executor_from_uops` easier? This fit in nicely, since we just call it on the new executor as it goes out the door.